src/runtime/profiler_common.cpp

```diff
@@ -91,60 +99,49 @@ WEAK halide_profiler_pipeline_stats *find_or_create_pipeline(const char *pipelin
         p->funcs[i].active_threads_numerator = 0;
         p->funcs[i].active_threads_denominator = 0;
     }
-    s->first_free_id += num_funcs;
     s->pipelines = p;
     return p;
 }
 
-WEAK void bill_func(halide_profiler_state *s, int func_id, uint64_t time, int active_threads) {
-    halide_profiler_pipeline_stats *p_prev = nullptr;
-    for (halide_profiler_pipeline_stats *p = s->pipelines; p;
-         p = (halide_profiler_pipeline_stats *)(p->next)) {
-        if (func_id >= p->first_func_id && func_id < p->first_func_id + p->num_funcs) {
-            if (p_prev) {
-                // Bubble the pipeline to the top to speed up future queries.
-                p_prev->next = (halide_profiler_pipeline_stats *)(p->next);
-                p->next = s->pipelines;
-                s->pipelines = p;
-            }
-            halide_profiler_func_stats *f = p->funcs + func_id - p->first_func_id;
-            f->time += time;
-            f->active_threads_numerator += active_threads;
-            f->active_threads_denominator += 1;
-            p->time += time;
-            p->samples++;
-            p->active_threads_numerator += active_threads;
-            p->active_threads_denominator += 1;
-            return;
-        }
-        p_prev = p;
-    }
-    // Someone must have called reset_state while a kernel was running. Do nothing.
+WEAK void update_running_instance(halide_profiler_instance_state *instance, uint64_t time) {
+    halide_profiler_func_stats *f = instance->funcs + instance->current_func;
+    f->time += time;
+    f->active_threads_numerator += instance->active_threads;
+    f->active_threads_denominator += 1;
+    instance->samples++;
+    instance->active_threads_numerator += instance->active_threads;
+    instance->active_threads_denominator += 1;
+    instance->time += time;
 }
 
 extern "C" WEAK int halide_profiler_sample(struct halide_profiler_state *s, uint64_t *prev_t) {
-    int func, active_threads;
+    if (!s->instances) {
+        // No Halide code is currently running
+        return 0;
+    }
+    halide_profiler_instance_state *instance = s->instances;
+
     if (s->get_remote_profiler_state) {
         // Execution has disappeared into remote code running
         // on an accelerator (e.g. Hexagon DSP)
-        s->get_remote_profiler_state(&func, &active_threads);
-    } else {
-        func = s->current_func;
-        active_threads = s->active_threads;
+
+        if (s->instances->next) {
+            // Unfortunately a user_context is not available or meaningful if
+            // there are zero or multiple instances running.
+            halide_error(nullptr, "Cannot use device profiling while multiple profiled pipelines are running on host. Shutting down profiler.");
+            return -1;
```

Comment:
I have never worked with this Hexagon thing, but perhaps shutting down the profiler is an overkill? Might be not user-friendly. You still might want to see the profiler results for another pipeline that is running alone on a later point in the program? Instead, omitting the profile report for all pipelines that have run simultaneously might be a less extreme alternative? As I said: no clue what I'm talking about, but trying to spot potential annoyances for anyone using this.

Reply:
I think this situation means that the profiler state is already corrupt. This should already have thrown an error in hexagon_host when trying to launch the remote kernel. If you try to launch a profiled hexagon kernel, and there's already an instance in flight, it will refuse to even launch the kernel. I'll change it to a hard assert because it's supposed to be unreachable.

Reply:
I changed it to a debug assert in this place, and instead put in a hard error message if you either try to launch a profiled pipeline while a profiled pipeline is running on device, or try to launch a profiled pipeline on device while another profiled pipeline is running.